bindings/rust/s2n-tls/src/callbacks.rs

```diff
@@ -118,30 +124,128 @@ pub(crate) fn trigger_async_callback<T: 'static + AsyncCallback>(
+/// [`Config`] on the connection.
+pub trait AsyncClientHelloFuture {
+    fn poll_client_hello(
+        self: Pin<&mut Self>,
+        connection: &mut Connection,
+        ctx: &mut core::task::Context,
+    ) -> Poll<Result<(), Error>>;
+}
+
+pin_project! {
+/// An implementation of [`AsyncClientHelloFuture`] which resolves the provided
+/// future and sets the config on the [`connection::Connection`].
+pub struct ConfigResolver<F: Future<Output = Result<Config, Error>>> {
+    #[pin]
+    fut: F,
+}
+}
+
+impl<F: Future<Output = Result<Config, Error>>> ConfigResolver<F> {
+    pub fn new(fut: F) -> Self {
+        ConfigResolver { fut }
+    }
+}
+
+impl<F: Future<Output = Result<Config, Error>>> AsyncClientHelloFuture for ConfigResolver<F> {
+    fn poll_client_hello(
+        self: Pin<&mut Self>,
+        connection: &mut Connection,
+        ctx: &mut core::task::Context,
+    ) -> Poll<Result<(), Error>> {
+        let this = self.project();
+        let config = match this.fut.poll(ctx) {
+            Poll::Ready(config) => config?,
+            Poll::Pending => return Poll::Pending,
+        };
+
+        connection.set_config(config)?;
+
+        Poll::Ready(Ok(()))
+    }
 }
 
 /// A trait for the callback executed after parsing the TLS Client Hello.
 ///
 /// Use in conjunction with
 /// [config::Builder::set_client_hello_callback](`crate::config::Builder::set_client_hello_callback()`).
 pub trait ClientHelloCallback {
-    fn poll_client_hello(&self, connection: &mut Connection) -> Poll<Result<(), Error>>;
+    /// The application can return a `None` to resolve the client_hello_callback
+    /// synchronously or return a `Some(AsyncClientHelloFuture)` if it wants to
+    /// run some asynchronous task beforing resolving the callback.
+    ///
+    /// [`ConfigResolver`], which implements [`AsyncClientHelloFuture`] can be
+    /// returned if the application wants to set a new [`Config`] on the connection.
+    ///
+    /// If the server_name is used to configure the connection then the application
+    /// should call [`connection::Connection::server_name_extension_used()`].
+    fn on_client_hello(
+        // this should NEVER be a `&mut` since its would be possible to
+        // mutate the Config on the Connection. Since one Config can be
+        // used for multiple Connections that would be undefined-behavior
+        &self,
+        connection: &mut Connection,
+    ) -> Option<Pin<Box<dyn AsyncClientHelloFuture>>>;
 }
 
 pub(crate) struct AsyncClientHelloCallback {}
 impl AsyncCallback for AsyncClientHelloCallback {
-    fn poll(&mut self, conn: &mut Connection) -> Poll<Result<(), Error>> {
-        let result = conn
+    fn poll(&self, conn: &mut Connection) -> Poll<Result<(), Error>> {
+        // if there is already a future set on the connection then take it and poll it
+        if let Some(mut fut) = conn.take_client_hello_future() {
+            let waker = conn.waker().ok_or(Error::CALLBACK_EXECUTION)?.clone();
```

Comment:
why do you need to clone? you're borrowing it in the `from_waker` function

Reply:
the function below `ConnectionFuture::poll` (after name change) takes a `&mut Connection, &mut Context` so we get borrow issues.

I thought about implementing a `with_connection` but that would be some dedupe code and didnt seem worth it since Waker should be cheap to clone.